### PR TITLE
Spawning, dynamic rooms, and dynamic walls

### PR DESCRIPTION
**Pull request by AleksandrHovhannisyan**
Spawn scripts and points added from the prototype (nearly identical, very minor changes).

Rooms and walls/doors are just about the same as they were in the prototype, except you can resize them dynamically by adjusting the settings in the editor for the room width/length, door width, and wall width. This will allow us to create some really interesting variations. For example, we could designate a room in the level as a boss room by giving it an especially wide door, for example.

Let me know if you have any suggestions for improvement.

### CODE REVIEW

**Review by Jazy5552 — CHANGES_REQUESTED** [review]
I don't think this is efficient for constructing rooms. I see where you are going with the whole `Door Width` and `Wall Width` but it just doesn't bold well when the editor shows the walls as one width, then at run-time they expend/contract to a different size. Generally these sort of things are done in the editor.

I would suggest we adopt a grid like system for constructing rooms. So we can have a segment that is just a wall, another is just a door, and... I guess that's it for now. The good thing about this is that we can have different sized rooms and even get creative with their structures (Not just squares/rectangles). Maybe even use the same grid style for the floors (So we can add hazards later perhaps). Due to the room logic being primarily triggered by a collider over the room floor, it may be difficult to find a solution for when the floor is split into pieces. But I think it can be done. 

Also how about we make the doors swing open, instead of rise upwards. Since this is a top down game, the rising door animation is not really visible, but a swinging door would be much more visible (Specially if the player could swing it open themselves).

Basically I'm saying to scrap this to make it more extendable than it was in the prototype. Please do feel free to discuss/argue this with me if you do think this is the best way (or there is a better way). If you'd prefer we can even hangouts to talk about some design/implementation ideas :)

**Comment by AleksandrHovhannisyan** [comment]
> it just doesn't bold well when the editor shows the walls as one width, then at run-time they expend/contract to a different size. Generally these sort of things are done in the editor.

Good point, I'l scrap this and figure something else out.

> So we can have a segment that is just a wall, another is just a door, and... I guess that's it for now. The good thing about this is that we can have different sized rooms and even get creative with their structures (Not just squares/rectangles).

This is a little easier said than done. We can't have a wall and a door separately unless we can mold the wall to accommodate a place for the door. If the wall is just going to have a predefined hole, then it doesn't make sense to not give it a door. One thought that comes to mind is maybe making each wall consist of door-sized blocks, and maybe we can delete one of those in the editor to make room for a door? Not sure.

> Also how about we make the doors swing open, instead of rise upwards. Since this is a top down game, the rising door animation is not really visible

It isn't visible now because the doors are the same color as the rest of the materials in the scene. If you recall the prototype, it was actually quite noticeable. I would prefer we stick with rising doors; they give off more of a dungeon vibe than ordinary doors that swing open.

> Due to the room logic being primarily triggered by a collider over the room floor, it may be difficult to find a solution for when the floor is split into pieces. But I think it can be done.

I don't see why we'd split the floor into pieces. Don't fix things that don't need fixing :)

**Comment by Jazy5552** [comment]
> One thought that comes to mind is maybe making each wall consist of door-sized blocks

Exactly what I'm thinking. Reason I'm suggesting this is so that we are not tied to just having rectangular/square rooms. If we use a grid system for room construction then we can have almost any shaped room. Which is why I suggested splitting the floor, that way the floor can follow the shape of the room.

**Review by Jazy5552 — APPROVED** [review]
This looks fine I suppose (Same stuff from the prototype it seems). I'm going to take over and work on the room implementation design as we discussed.
